### PR TITLE
[SSHD-1033]: Fix use DynamicPortForwarding + LocalPortForwarding

### PR DESCRIPTION
**Pull request by ggrandes**
https://issues.apache.org/jira/browse/SSHD-1033

### CODE REVIEW

**Comment by ggrandes** [comment]
I don't understand the fail, I checked in local with jdk 1.8 and 13

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java [hunk 1]
```diff
@@ -974,22 +975,36 @@ protected void preClose() {
...
+	}
+
+	private IoAcceptor getLocalIoAcceptor(Factory<? extends IoHandler> handlerFactory) {
+		if (localAcceptor == null) {
+			localAcceptor = createIoAcceptor(handlerFactory);
+		}
+		return localAcceptor;
+	}
+
+	private IoAcceptor getDynamicIoAcceptor(Factory<? extends IoHandler> handlerFactory) {
```
Comment (lgoldstein) [inline sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:993]: Let's make these `protected` in accordance with our policy to enable users to easily override our implementation if they see fit.

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java [hunk 2]
```diff
@@ -964,7 +965,7 @@ protected void invokePortEventListenerSignallerHolders(
 
     @Override
     protected synchronized Closeable getInnerCloseable() {
-        return builder().parallel(toString(), dynamicLocal.values()).close(acceptor).build();
+        return builder().parallel(toString(), dynamicLocal.values()).close(dynamicAcceptor).build();
```
Comment (lgoldstein) [inline sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:968]: Don't we need to close the local acceptor as well ?

**Comment by lgoldstein** [comment]
Remember to also update the `CHANGES.md` file about this bugfi

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 2 above)
Comment (ggrandes) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:968]: It is possible, but honestly, I don't know the sshd code enough and I just let myself be guided by that "dynamicLocal.values()"

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 1 above)
Comment (ggrandes) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:993]: The original caller doBind is private too; must be changed to protected too to facilitate this overriding?

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 2 above)
Comment (gnodet) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:968]: Yes, both would have to be closed.

**Comment by gnodet** [comment]
I would like a proper unit test for that.  I've investigated the issue and wrote one, so it should be included in this PR too.  I'll provide a link tomorrow.  However, I spent a bit of time looking at the issue and I haven't yet understood what the exact problem is and why using two acceptors solves the problem.

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 2 above)
Comment (ggrandes) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:968]: Well, but... I don't known how to do... '':-)

**Comment by gnodet** [comment]
Here is a proper unit test:
  https://github.com/gnodet/mina-sshd/blob/SSHD-1033/sshd-core/src/test/java/org/apache/sshd/common/forward/Sshd1033Test.java

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 2 above)
Comment (gnodet) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:968]:
```
        return builder().parallel(toString(), dynamicLocal.values())
                .close(localAcceptor).close(dynamicAcceptor).build();
```

**Comment by gnodet** [comment]
Ok, I understand the fix now.  I'm ok with it.

Please include the unit test I pointed to, add the required line to the `CHANGES.md`, fix the `getInnerCloseable()` method and squash the commits into a single one.

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 2 above)
Comment (ggrandes) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:968]: A lot of thanks!

**Comment by ggrandes** [comment]
@gnodet ok, is my first PR, what need to put in CHANGES.md? some example?

**Comment by gnodet** [comment]
> @gnodet ok, is my first PR, what need to put in CHANGES.md? some example?

https://github.com/gnodet/mina-sshd/commit/2057f9734dc9fb2a1ca03d3d8c73ff4c4ba2d3c4

File: sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java (hunk 1 above)
Comment (gnodet) [reply sshd-core/src/main/java/org/apache/sshd/common/forward/DefaultForwardingFilter.java:993]: > The original caller `doBind` is `private` too; must be changed to `protected` too to facilitate this overriding?

Yes

**Comment by ggrandes** [comment]
@gnodet [code](https://github.com/ggrandes-clones/mina-sshd/tree/hotfix/SSHD-1033) done, but squash the commits (needed push -f) seem has broken the pull-request (I only see old commits in PR, but no new), you know how fix this?

**Comment by gnodet** [comment]
> @gnodet [code](https://github.com/ggrandes-clones/mina-sshd/tree/hotfix/SSHD-1033) done, but squash the commits (needed push -f) seem has broken the pull-request (I only see old commits in PR, but no new), you know how fix this?

Right, you may not have the karma do force push a PR.  I'll have a look at your branch then.

**Comment by gnodet** [comment]
Merged in https://github.com/apache/mina-sshd/commit/fb0b0c4dfac7df2f5a4102d87397e12171615fbb, thx a ton !

**Comment by ggrandes** [comment]
@lgoldstein @gnodet Thanks to both

**Pull request closed without merging**